### DISCUSSION TITLE
Why does pyright deduce Mapping[Unknown, Unknown] in this case?

### DISCUSSION THREAD

**Discussion by Holt59** [microsoft/pyright#7861] [Q&A] [answered]
I have the following function that I would like to use to convert either a mapping or a float to a mapping (with `defaultdict`):

```python
def asdict(value: Mapping[_K, _V] | _V) -> Mapping[_K, _V]:
    if not isinstance(value, Mapping):
        return defaultdict(lambda: value)
    return value
```

I get a warning from `pyright` because in the second `return`, the type of `value` is deduced as `Mapping[_K@asdict, _V@asdict] | Mapping[Unknown, Unknown]*`.

Why is the type deduce like that? Is there a way to have it deduce `Mapping[_K, _V]`? If no, can this be safely ignored?

**Comment by erictraut** [comment, marked as answer]
I presume that you have strict mode enabled because pyright does not generate any type errors for this code in standard mode.

I'm guessing that `_K` and `_V` are type variables with no explicit upper bound, in which case they have implicit upper bounds of `object`. I can't tell from your code snippet wither `_K` and `_V` are scoped to a class. I'm guessing that they're not.

A simplified version of the function should make it more clear why pyright narrows `_V` to `Mapping[Unknown, Unknown]` in this case.

```python
def func[_V](value: _V):
    if isinstance(value, Mapping):
        reveal_type(value) # Mapping[Unknown, Unknown]
```

Here `value` could be of any type including a `Mapping`. The type arguments are unknown, so pyright must assume that the type is `Mapping[Unknown, Unknown]`.

**Reply by Holt59** [reply]
Thanks for the clarification, and your hypothesis are right (`_K` and `_V` have no bounds and are no scoped).

If I summarize, the issue is because `_V` could be a mapping type itself?
